node/pkg/tss/implementation.go

```diff
@@ -0,0 +1,381 @@
+
+			// todo: ensure someone listens to this channel.
+			//todo: wrap the message into a gossip message and output it to the network, sign (or encrypt and mac) it and send it.
+			t.gossipOutChan <- tssMsg
+			// fmt.Printf("guardian %v sent %v \n", t.GuardianStorage.Self.Index, m.Type())
+		case err := <-t.fpErrChannel:
+			_ = err // todo: log the error?
+			// fmt.Printf("guardian %v received error %v \n", t.GuardianStorage.Self.Index, err)
+		}
+	}
+}
+
+func protoToPartyId(pid *gossipv1.PartyId) *tss.PartyID {
+	return &tss.PartyID{
+		MessageWrapper_PartyID: &tss.MessageWrapper_PartyID{
+			Id:      pid.Id,
+			Moniker: pid.Moniker,
+			Key:     pid.Key,
+		},
+		Index: int(pid.Index),
+	}
+}
+func partyIdToProto(pid *tss.PartyID) *gossipv1.PartyId {
+	return &gossipv1.PartyId{
+		Id:      pid.Id,
+		Moniker: pid.Moniker,
+		Key:     pid.Key,
+		Index:   uint32(pid.Index),
+	}
+}
+func (t *Engine) intoGossipMessage(m tss.Message) (*gossipv1.GossipMessage, error) {
+	bts, routing, err := m.WireBytes()
+	if err != nil {
+		return nil, err
+	}
+
+	indices := make([]*gossipv1.PartyId, 0, len(routing.To))
+	for _, pId := range routing.To {
+		indices = append(indices, partyIdToProto(pId))
+	}
+
+	msgToSend := &gossipv1.SignedMessage{
+		Payload:         bts,
+		Sender:          partyIdToProto(m.GetFrom()),
+		Recipients:      indices,
+		MsgSerialNumber: atomic.AddUint64(&t.msgSerialNumber, 1),
+		Authentication:  nil,
+	}
+
+	tssMsg := &gossipv1.PropagatedMessage{}
+
+	if routing.IsBroadcast || len(routing.To) == 0 || len(routing.To) > 1 {
+		t.sign(msgToSend)
+		tssMsg.Payload = &gossipv1.PropagatedMessage_Echo{
+			Echo: &gossipv1.Echo{
+				Message:   msgToSend,
+				Signature: nil, //  No sig here, it means this is the original sender of the message, and not a vote.
+				Echoer:    0,   // TODO: use -1 since this is not an echo.
+			},
+		}
+	} else {
+		t.encryptAndMac(msgToSend)
+		// encrypt then mac the msgToSend (payload encrypted, and mac the full &gossipv1.SignedMessage)
+		tssMsg.Payload = &gossipv1.PropagatedMessage_Unicast{
+			Unicast: msgToSend,
+		}
+	}
+
+	return &gossipv1.GossipMessage{
+		Message: &gossipv1.GossipMessage_TssMessage{
+			TssMessage: tssMsg,
+		},
+	}, nil
+}
+
+func (t *Engine) HandleIncomingTssMessage(msg *gossipv1.GossipMessage_TssMessage) {
+	if t == nil {
+		return
+	}
+
+	defer func() {
+		// todo: consider sending the message to be gossiped or not. perhaps it's a duplicate message?
+	}()
+
+	switch m := msg.TssMessage.Payload.(type) {
+	case *gossipv1.PropagatedMessage_Unicast:
+		parsed, err := t.handleUnicast(m)
+		if err != nil {
+			return
+		}
+
+		// TODO: add the uuid of this message to the set of received messages.
+		t.fp.Update(parsed)
+	case *gossipv1.PropagatedMessage_Echo:
+		parsed, err := t.handleEcho(m)
+		if err != nil {
+			return
+		}
+
+		// fmt.Printf("guardian %v received from %v: %v\n", t.GuardianStorage.Self.Index, parsed.GetFrom().Index, parsed.Type())
+		t.fp.Update(parsed)
```

Comment:
nit: t.fp.Update(parsed) is common to both cases, so you could do it out of the switch

Reply:
I kept it like that since there should be a difference once we introduce the reliable broadcast.
The changes between the branches are present in the `relbroad` branch.